plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/maven2/metadata/MetadataRebuilder.java

```diff
@@ -0,0 +1,418 @@
+        }
+      }
+    }
+    sql.append(" GROUP BY group, name LIMIT=-1");
+    final Worker worker = new Worker(repository, update, sql.toString(), sqlParams);
+    try (StorageTx tx = repository.facet(StorageFacet.class).openTx()) {
+      final ORID bucketOrid = bucketEntityAdapter.decode(tx.getBucket().getEntityMetadata().getId());
+      sqlParams.put("bucket", bucketOrid);
+      worker.rebuild(tx);
+    }
+  }
+
+  /**
+   * Inner class (non static) that encapsulates the work, as metadata builder is stateful.
+   */
+  private class Worker
+  {
+    private final boolean update;
+
+    private final String sql;
+
+    private final Map<String, Object> sqlParams;
+
+    private final StorageFacet storageFacet;
+
+    private final MavenFacet mavenFacet;
+
+    private final MavenPathParser mavenPathParser;
+
+    private final MetadataBuilder metadataBuilder;
+
+    private final MetadataUpdater metadataUpdater;
+
+    private final DocumentBuilderFactory documentBuilderFactory;
+
+    public Worker(final Repository repository,
+                  final boolean update,
+                  final String sql,
+                  final Map<String, Object> sqlParams)
+    {
+      this.update = update;
+      this.sql = sql;
+      this.sqlParams = sqlParams;
+      this.storageFacet = repository.facet(StorageFacet.class);
+      this.mavenFacet = repository.facet(MavenFacet.class);
+      this.mavenPathParser = mavenFacet.getMavenPathParser();
+      this.metadataBuilder = new MetadataBuilder();
+      this.metadataUpdater = new MetadataUpdater(mavenFacet);
+      this.documentBuilderFactory = DocumentBuilderFactory.newInstance();
+    }
+
+    public void rebuild(final StorageTx tx)
+    {
+      final ODatabaseDocumentTx readDatabase = tx.getDb();
+      readDatabase.command(
+          new OSQLAsynchQuery<ODocument>(
+              sql,
+              new OCommandResultListener()
+              {
+                String currentGroupId = null;
+
+                @Override
+                public boolean result(Object iRecord) {
+                  try {
+                    final ODocument doc = (ODocument) iRecord;
+                    final String groupId = doc.field("groupId", OType.STRING);
+                    final String artifactId = doc.field("artifactId", OType.STRING);
+                    final Set<String> baseVersions = doc.field("baseVersions", OType.EMBEDDEDSET);
+
+                    final boolean groupChange = !Objects.equals(currentGroupId, groupId);
+                    if (groupChange) {
+                      if (currentGroupId != null) {
+                        processMetadata(metadataMavenPath(currentGroupId, null, null), metadataBuilder.onExitGroupId());
+                      }
+                      currentGroupId = groupId;
+                      metadataBuilder.onEnterGroupId(groupId);
+                    }
+                    rebuildAbV((ORID) sqlParams.get("bucket"), groupId, artifactId, baseVersions);
```

Comment:
So I'm curious as to how this will work when we start working with remote databases. The doco for the asynch query states: 
```NOTE: if you're working with remote databases don't
 * execute any remote call inside the callback function because the network channel is locked until the query command has finished.```
and this in turn makes calls using a different part of the orient API that does not (at least immediately) appear to be async and could potentially block. Might be I'm overcomplificating the scenario at this early stage, but we should probably be keeping our eyes open to ensure we're consistent in our usage and don't accidentally introduce things we'll just need to unravel with planned features.